kythe/go/extractors/cmd/rust/rust_project_to_kzip.go

```diff
@@ -0,0 +1,552 @@
+	// We need to write two sets of files to the kzip:
+	// 1. the set of files in this crate, as Source files
+	// 2. the other required inputs for this crate which aren't in the set of source files for this crate (dependencies)
+	// We can use collectCrateSources for both, with a different argument of source dirs
+	var crateFiles []string
+	var crateRequiredInputs []*apb.CompilationUnit_FileInput
+	var err error
+	if crateFiles, crateRequiredInputs, err = collectCrateSources(ctx, e, sourceDirs[crate.CrateId]); err != nil {
+		fmt.Printf("Error getting source files for crate %s: %v\n", crate.Label, err)
+		return err
+	}
+
+	// Get the set of depended-on files for the entire crate, not just direct source files
+	var requiredInputs []*apb.CompilationUnit_FileInput = crateRequiredInputs
+	for _, dep := range transitiveDeps[crate.CrateId] {
+		_, depRequiredInputs, err := collectCrateSources(ctx, e, sourceDirs[dep])
+		if err != nil {
+			fmt.Printf("Error getting source files for crate %s: %v\n", crate.Label, err)
+			return err
+		}
+		requiredInputs = append(requiredInputs, depRequiredInputs...)
+	}
+
+	// Synthesize a rust-project.json for this crate and add it to the CU
+	projectJsonDigest, err := e.kzipWriter.AddFile(getProjectBytes(synthesizeProjectJson(crate.CrateId, e.project.Crates, transitiveDeps)))
+	if err != nil {
+		fmt.Printf("Error adding project json to kzip: %v\n", err)
+		return err
+	}
+	projectJsonVname := &spb.VName{
+		Corpus:   e.corpus,
+		Language: "json",
+		Path:     "rust-project.json",
+	}
+	requiredInputs = append(requiredInputs, &apb.CompilationUnit_FileInput{
+		VName: projectJsonVname,
+		Info: &apb.FileInfo{
+			Path:   "rust-project.json",
+			Digest: projectJsonDigest,
+		},
+	})
+
+	compilationUnit := &apb.CompilationUnit{
+		VName: &spb.VName{
+			Corpus:   e.corpus,
+			Language: "rust",
+		},
+		RequiredInput: requiredInputs,
+		SourceFile:    crateFiles,
+		Argument:      crate.CompilerArgs,
+	}
+
+	digest, err := e.kzipWriter.AddUnit(compilationUnit, nil)
+	if err != nil {
+		fmt.Printf("Error adding compilation unit to kzip: %v, crate %s, digest: %s\n", err, crate.Label, digest)
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	projectJSONPath := flag.String("project_json", "", "Path to the rust-project.json file (required)")
+	outputZipPath := flag.String("output", "", "Path for the output kzip file (required)")
+	projectRoot := flag.String("root", "", "Directory to which all paths to source files are relative; the root directory to all possible source files. (required)")
+	corpus := flag.String("corpus", "", "Corpus name for VNames (required)")
+	vnamesJsonPath := flag.String("vnames_json_path", "", "Path to vnames.json (required)")
+	crateFilter := flag.Int("crate_filter", 0, "optional, a way to filter the output to a specific crate ID")
```

Comment:
nit: "a way to filter" is vague - if you'd like to keep this flag, be more specific?

(I'd suggest that "path to root module" might be a bit easier to describe and to use)